lib/linguist/heuristics.yml

```
@@ -136,6 +136,10 @@ disambiguations:
   rules:
   - language: XML
     pattern: '^(\s*)(?i:<Project|<Import|<Property|<?xml|xmlns)'
+- extensions: ['.cairo']
+  rules:
+    - language: Cairo 0
+      pattern: '(^(\s*)%lang(\s+)([A-Za-z0-9_]+))|(^(\s*)%builtins(\s+)([A-Za-z0-9_]+\s*)*$)|(^(\s*)from(\s+)starkware\.(cairo|starknet)\.([A-Za-z0-9_.\s]+?)import)'
```

As these are the only two languages using this extension, we can force things to Cairo here rather than taking things further into the classifier:
```suggestion
      pattern: '(^(\s*)%lang(\s+)([A-Za-z0-9_]+))|(^(\s*)%builtins(\s+)([A-Za-z0-9_]+\s*)*$)|(^(\s*)from(\s+)starkware\.(cairo|starknet)\.([A-Za-z0-9_.\s]+?)import)'
    - language: Cairo
```

---

I thought about this, but I would have to extend this heuristic to match [CASM](https://github.com/github-linguist/linguist/blob/066b07b201e9ac1b287ed0560de57f406e1541a3/samples/Cairo%200/fib_u128.cairo) files, which would blow up this already complex regex. 😕 I want CASM to be counted as Cairo 0 (it's a strict subset), yet they do not exhibit such trivial and distinguishable constructs I detect here.

---

I removed this heuristic. With new samples, the classifier now works great.

---

I'm happy with this, but Cairo user may not be 😁  

The classifier is not very reliable and is really a "last resort" - it's literally the last thing we try if we can't find a result any other way. You might want to consider a simple heuristic that would match more often than not to help things and leave the classifier to those occasions where the heuristic doesn't match. There's no need to be 100% perfect; close enough is good enough. 

Maybe try matching for most of Cairo, like your search result, and leave the classifier to handle Cairo 0 and CASM and the more obscure uses?

---

Done. Went with your original suggestion to try to match most of `Cairo 0` and fall back to `Cairo` because I realised there is one more very common pattern in `Cairo 0` code. Now, I believe one would have to handcraft Cairo 0 code deliberately to be identified as Cairo.